### PR TITLE
Add intermediate progress markers for xptracker plugin

### PR DESCRIPTION
**Pull request by JAnken123**
Modified progressbars to include intermediate progress markers and used it in the xptracker plugin to show intermediate levels per issue #6863. Defaults to not enabled.

Looks like this:
image

Edit: Current version looks like this:
image

### CODE REVIEW

**Comment by deathbeam** [comment]
Can the display be made a bit less... extreme?

**Comment by JAnken123** [comment]
Would any of these be viable?

image

image

**Comment by deathbeam** [comment]
1st one looks good + make it slimmer and more dull

**Comment by forsco** [comment]
If the bottom one is used however i wouldn't display levels already surpassed.

**Comment by JAnken123** [comment]
I made it a bit darker. If going forsco's way it would look like pic2, which I think is more aesthetically pleasing but has less information.

image

image

**Comment by deathbeam** [comment]
No, because levels already surpassed are useless information and it looks very uggly.

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 1]
```diff
@@ -207,6 +211,23 @@ private void rebuildAsync(boolean updated, boolean skillPaused, XpSnapshotSingle
 				? "200M"
 				: "Lvl. " + xpSnapshotSingle.getEndLevel());
 
+			// Add intermediate level positions to progressBar
+			ArrayList<Double> positions = new ArrayList<>();
+			if (xpTrackerPlugin.getXpTrackerConfig().showIntermediateLevels() && xpSnapshotSingle.getEndLevel() - xpSnapshotSingle.getStartLevel() > 1)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: You can simply inject xpTrackerConfig, or send it via ctor here, remove the xpTrackerPlugin call

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerConfig.java [hunk 2]
```diff
@@ -54,7 +54,18 @@ default boolean pauseOnLogout()
 	}
 
 	@ConfigItem(
-		position = 2,
+			position = 2,
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerConfig.java]: Fix your indentation

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerPlugin.java [hunk 3]
```diff
@@ -87,6 +88,7 @@
 	@Inject
 	private SkillIconManager skillIconManager;
 
+	@Getter
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpTrackerPlugin.java]: Revert this

File: runelite-client/src/main/java/net/runelite/client/ui/components/ProgressBar.java [hunk 4]
```diff
@@ -28,9 +28,11 @@
 import java.awt.Color;
 import java.awt.Dimension;
 import java.awt.Graphics;
+import java.util.ArrayList;
 import javax.swing.JLabel;
 import javax.swing.SwingConstants;
 import javax.swing.border.EmptyBorder;
+
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/components/ProgressBar.java]: Remove this empty line

File: runelite-client/src/main/java/net/runelite/client/ui/components/ProgressBar.java [hunk 5]
```diff
@@ -52,6 +54,9 @@
 	private String centerLabelText = "";
 	private String dimmedText = "";
 
+	@Setter
+	private ArrayList<Double> intermediateProgressPositions = new ArrayList<>();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/components/ProgressBar.java]: Move this next to other setters

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 6]
```diff
@@ -90,8 +91,11 @@
 
 	private boolean paused = false;
 
-	XpInfoBox(XpTrackerPlugin xpTrackerPlugin, Client client, JPanel panel, Skill skill, SkillIconManager iconManager) throws IOException
+	private XpTrackerConfig xpTrackerConfig;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]: Mark this final, move it above paused

File: runelite-client/src/main/java/net/runelite/client/ui/components/ProgressBar.java [hunk 7]
```diff
@@ -46,6 +47,9 @@
 	@Setter
 	private int value;
 
+	@Setter
+	private ArrayList<Double> intermediateProgressPositions = new ArrayList<>();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/components/ProgressBar.java]:
```suggestion
	private List<Double> positions = new ArrayList<>();
```

File: runelite-client/src/main/java/net/runelite/client/ui/components/ProgressBar.java [hunk 8]
```diff
@@ -89,6 +93,17 @@ public void paint(Graphics g)
 		super.paint(g);
 		g.setColor(getForeground());
 		g.fillRect(0, 0, topWidth, 16);
+
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/components/ProgressBar.java]:
```suggestion
```

File: runelite-client/src/main/java/net/runelite/client/ui/components/ProgressBar.java [hunk 9]
```diff
@@ -89,6 +93,17 @@ public void paint(Graphics g)
...
+		g.setColor(getForeground().darker());
+		for (Double position : intermediateProgressPositions)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/components/ProgressBar.java]:
```suggestion

		for (final Double position : positions)
```

File: runelite-client/src/main/java/net/runelite/client/ui/components/ProgressBar.java [hunk 10]
```diff
@@ -89,6 +93,17 @@ public void paint(Graphics g)
...
+		{
+			int xCord = (int) (getSize().width * (position));
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/components/ProgressBar.java]:
```suggestion
			final int xCord = (int) (getSize().width * (position));
```

File: runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java [hunk 11]
```diff
@@ -207,6 +211,23 @@ private void rebuildAsync(boolean updated, boolean skillPaused, XpSnapshotSingle
...
+				{
+					double relativeStartExperience = Experience.getXpForLevel(level) - xpSnapshotSingle.getStartGoalXp();
+					double relativeEndExperience = xpSnapshotSingle.getEndGoalXp() - xpSnapshotSingle.getStartGoalXp();
+					positions.add(relativeStartExperience / relativeEndExperience);
+				}
+				progressBar.setIntermediateProgressPositions(positions);
+			}
+			else
+			{
+				progressBar.setIntermediateProgressPositions(positions);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java]:
```suggestion
				progressBar.setPositions(positions);
```

**Comment by deathbeam** [comment]
Any update @JAnken123 ?

**Comment by JAnken123** [comment]
I am unwell. I will look at this as soon as I get healthy again.

**Pull request merged**